File: mobile/src/main/java/net/mm2d/dmsexplorer/view/adapter/PropertyAdapter.java
/*
 * Copyright (c) 2017 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.dmsexplorer.view.adapter;

import android.content.Context;
import android.databinding.DataBindingUtil;
import android.graphics.Paint;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.RecyclerView.Adapter;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import net.mm2d.dmsexplorer.R;
import net.mm2d.dmsexplorer.Repository;
import net.mm2d.dmsexplorer.databinding.PropertyListItemBinding;
import net.mm2d.dmsexplorer.view.adapter.PropertyAdapter.ViewHolder;
import net.mm2d.dmsexplorer.viewmodel.PropertyItemModel;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * 詳細情報の各項目をRecyclerViewを使用して表示するためのAdapter。
 *
 * @author <a href="mailto:ryo@mm2d.net">大前良介(OHMAE Ryosuke)</a>
 */
public class PropertyAdapter extends Adapter<ViewHolder> {
    private static final Pattern URL_PATTERN =
            Pattern.compile("https?://[\\w/:%#$&?()~.=+\\-]+");

    public enum Type {
        TITLE,
        TEXT,
        LINK,
        DESCRIPTION
    }

    public static class Entry {
        private final String mName;
        private final String mValue;
        private final Type mType;

        private Entry(
                @NonNull final String name,
                @NonNull final String value,
                @NonNull final Type type) {
            mName = name;
            mValue = value;
            mType = type;
        }

        @NonNull
        public String getName() {
            return mName;
        }

        @NonNull
        public String getValue() {
            return mValue;
        }

        @NonNull
        public Type getType() {
            return mType;
        }
    }

    private final Context mContext;
    private final List<Entry> mList;
    private final LayoutInflater mInflater;

    public PropertyAdapter(@NonNull final Context context) {
        mContext = context;
        mInflater = LayoutInflater.from(context);
        mList = new ArrayList<>();
    }

    protected Context getContext() {
        return mContext;
    }

    public void addEntry(
            @NonNull final String name,
            @Nullable final String value) {
        addEntry(name, value, Type.TEXT);
    }

    public void addEntry(
            @NonNull final String name,
            @Nullable final String value,
            @NonNull final Type type) {
        if (value == null || value.isEmpty()) {
            return;
        }
        mList.add(new Entry(name, value, type));
    }

    public void addTitleEntry(@NonNull final String name) {
        mList.add(new Entry(name, "", Type.TITLE));
    }

    @Override
    public int getItemViewType(final int position) {
        return mList.get(position).getType() != Type.DESCRIPTION ? 0 : 1;
    }

    @Override
    public ViewHolder onCreateViewHolder(
            @NonNull final ViewGroup parent,
            int viewType) {
        return new ViewHolder(getContext(), mInflater,
                DataBindingUtil.inflate(mInflater, R.layout.property_list_item, parent, false));
    }

    @Override
    public void onBindViewHolder(
            @NonNull final ViewHolder holder,
            int position) {
        holder.applyItem(mList.get(position));
    }

    @Override
    public int getItemCount() {
        return mList.size();
    }

    static class ViewHolder extends RecyclerView.ViewHolder implements OnClickListener {
        private final Context mContext;
        private final LayoutInflater mInflater;
        private final PropertyListItemBinding mBinding;

        ViewHolder(
                @NonNull final Context context,
                @NonNull final LayoutInflater inflater,
                @NonNull final PropertyListItemBinding binding) {
            super(binding.getRoot());
            mContext = context;
            mInflater = inflater;
            mBinding = binding;
        }

        void applyItem(@NonNull final Entry entry) {
            String value = entry.getValue();
            if (entry.getType() == Type.DESCRIPTION) {
                value = setUpDescription(entry);
            }
            mBinding.setModel(new PropertyItemModel(entry.getName(), entry.getType(), value, this));
            mBinding.executePendingBindings();
        }

        private String setUpDescription(@NonNull final Entry entry) {
            final LinearLayout layout = mBinding.container;
            final int count = layout.getChildCount();
            for (int i = count - 1; i >= 2; i--) {
                layout.removeViewAt(i);
            }
            return makeDescription(layout, entry.getValue());
        }

        private String makeDescription(
                @NonNull final ViewGroup parent,
                @NonNull final String text) {
            String firstNormalText = "";
            final Matcher matcher = URL_PATTERN.matcher(text);
            int lastEnd = 0;
            while (matcher.find()) {
                final int start = matcher.start();
                final int end = matcher.end();
                if (start != lastEnd) {
                    final String normalText = trim(text.substring(lastEnd, start));
                    if (!normalText.isEmpty()) {
                        if (lastEnd == 0) {
                            firstNormalText = normalText;
                        } else {
                            addNormalText(parent, normalText);
                        }
                    }
                }
                addLinkText(parent, trim(text.substring(start, end)));
                lastEnd = end;
            }
            if (lastEnd == 0) {
                return trim(text);
            } else if (lastEnd != text.length()) {
                final String normalText = trim(text.substring(lastEnd, text.length()));
                if (!normalText.isEmpty()) {
                    addNormalText(parent, normalText);
                }
            }
            return firstNormalText;
        }

        private void addNormalText(
                @NonNull final ViewGroup parent,
                @NonNull final String text) {
            final TextView normal = (TextView) mInflater.inflate(R.layout.normal_text_view, parent, false);
            normal.setText(text);
            parent.addView(normal);
        }

        private void addLinkText(
                @NonNull final ViewGroup parent,
                @NonNull final String text) {
            final TextView link = (TextView) mInflater.inflate(R.layout.link_text_view, parent, false);
            link.setText(text);
            link.setPaintFlags(link.getPaintFlags() | Paint.UNDERLINE_TEXT_FLAG);
            link.setOnClickListener(this);
            parent.addView(link);
        }

        @Override
        public void onClick(final View v) {
            if (!(v instanceof TextView)) {
                return;
            }
            final CharSequence text = ((TextView) v).getText();
            if (TextUtils.isEmpty(text)) {
                return;
            }
            Repository.get().getOpenUriModel().openUri(mContext, text.toString());
        }
    }

    private static String trim(String str) {
        int len = str.length();
        int st = 0;
        while ((st < len) && isSpace(str.charAt(st))) {
            st++;
        }
        while ((st < len) && isSpace(str.charAt(len - 1))) {
            len--;
        }
        return ((st > 0) || (len < str.length())) ? str.substring(st, len) : str;
    }

    private static boolean isSpace(final char c) {
        return c <= '\u0020' || c == '\u00A0' || c == '\u3000';
    }
}

File: mobile/src/main/java/net/mm2d/dmsexplorer/view/adapter/PropertyAdapter.kt
/*
 * Copyright (c) 2017 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.dmsexplorer.view.adapter

import android.content.Context
import android.graphics.Paint
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnClickListener
import android.view.ViewGroup
import android.widget.TextView
import androidx.core.view.isVisible
import androidx.core.view.updatePadding
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.Adapter
import net.mm2d.android.upnp.cds.MediaServer
import net.mm2d.dmsexplorer.R
import net.mm2d.dmsexplorer.Repository
import net.mm2d.dmsexplorer.databinding.PropertyListItemBinding
import net.mm2d.dmsexplorer.domain.entity.ContentEntity
import net.mm2d.dmsexplorer.view.adapter.PropertyAdapter.ViewHolder
import net.mm2d.dmsexplorer.viewmodel.PropertyItemModel
import net.mm2d.dmsexplorer.viewmodel.adapter.TextViewBindingAdapter
import java.util.regex.Pattern

/**
 * 詳細情報の各項目をRecyclerViewを使用して表示するためのAdapter。
 *
 * @author [大前良介 (OHMAE Ryosuke)](mailto:ryo@mm2d.net)
 */
abstract class PropertyAdapter(
    protected val context: Context,
) : Adapter<ViewHolder>() {
    private val list: MutableList<Entry> = ArrayList()
    private val inflater: LayoutInflater = LayoutInflater.from(context)

    fun addEntry(
        name: String,
        value: String?,
        type: Type = Type.TEXT,
    ) {
        if (value.isNullOrEmpty()) {
            return
        }
        list.add(Entry(name, value, type))
    }

    fun addTitleEntry(
        name: String,
    ) {
        list.add(Entry(name, "", Type.TITLE))
    }

    override fun getItemViewType(
        position: Int,
    ): Int = if (list[position].type != Type.DESCRIPTION) 0 else 1

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int,
    ): ViewHolder =
        ViewHolder(
            context,
            inflater,
            PropertyListItemBinding.inflate(inflater, parent, false),
        )

    override fun onBindViewHolder(
        holder: ViewHolder,
        position: Int,
    ) {
        holder.applyItem(list[position])
    }

    override fun getItemCount(): Int = list.size

    enum class Type {
        TITLE,
        TEXT,
        LINK,
        DESCRIPTION,
    }

    class Entry(
        val name: String,
        val value: String,
        val type: Type,
    )

    class ViewHolder(
        private val context: Context,
        private val inflater: LayoutInflater,
        private val binding: PropertyListItemBinding,
    ) : RecyclerView.ViewHolder(binding.root),
        OnClickListener {

        fun applyItem(
            entry: Entry,
        ) {
            var value = entry.value
            if (entry.type == Type.DESCRIPTION) {
                value = setUpDescription(entry)
            }
            val model = PropertyItemModel(entry.name, entry.type, value, this)
            val margin = context.resources.getDimensionPixelSize(R.dimen.property_margin_bottom)
            binding.root.updatePadding(bottom = if (model.enableDescription) margin else 0)
            binding.title.text = model.title
            binding.description.text = model.description
            val marginLink = context.resources.getDimensionPixelSize(R.dimen.property_margin_link)
            val marginNormal = context.resources.getDimensionPixelSize(R.dimen.property_margin_normal)
            binding.description.updatePadding(
                top = if (model.isLink) marginLink else marginNormal,
                bottom = if (model.isLink) marginLink else marginNormal,
            )
            if (model.isLink) {
                binding.description.setBackgroundResource(R.drawable.bg_link_text)
            } else {
                binding.description.background = null
            }
            binding.description.isVisible = model.enableDescription
            binding.description.setTextIsSelectable(!model.isLink)
            binding.description.setOnClickListener(model.onClickListener)
            TextViewBindingAdapter.setTextColorAttr(
                binding.description,
                if (model.isLink) R.attr.themeLinkColor else R.attr.themeTextColor,
            )
            TextViewBindingAdapter.setUnderlineFlag(binding.description, model.isLink)
        }

        private fun setUpDescription(
            entry: Entry,
        ): String {
            val layout = binding.container
            val count = layout.childCount
            for (i in count - 1 downTo 2) {
                layout.removeViewAt(i)
            }
            return makeDescription(layout, entry.value)
        }

        private fun makeDescription(
            parent: ViewGroup,
            text: String,
        ): String {
            var firstNormalText = ""
            val matcher = URL_PATTERN.matcher(text)
            var lastEnd = 0
            while (matcher.find()) {
                val start = matcher.start()
                val end = matcher.end()
                if (start != lastEnd) {
                    val normalText = text.substring(lastEnd, start).trim { isSpace(it) }
                    if (normalText.isNotEmpty()) {
                        if (lastEnd == 0) {
                            firstNormalText = normalText
                        } else {
                            addNormalText(parent, normalText)
                        }
                    }
                }
                addLinkText(parent, text.substring(start, end).trim { isSpace(it) })
                lastEnd = end
            }
            if (lastEnd == 0) {
                return text.trim { isSpace(it) }
            } else if (lastEnd != text.length) {
                val normalText = text.substring(lastEnd, text.length).trim { isSpace(it) }
                if (normalText.isNotEmpty()) {
                    addNormalText(parent, normalText)
                }
            }
            return firstNormalText
        }

        private fun addNormalText(
            parent: ViewGroup,
            text: String,
        ) {
            val normal = inflater.inflate(R.layout.normal_text_view, parent, false) as TextView
            normal.text = text
            parent.addView(normal)
        }

        private fun addLinkText(
            parent: ViewGroup,
            text: String,
        ) {
            val link = inflater.inflate(R.layout.link_text_view, parent, false) as TextView
            link.text = text
            link.paintFlags = link.paintFlags or Paint.UNDERLINE_TEXT_FLAG
            link.setOnClickListener(this)
            parent.addView(link)
        }

        override fun onClick(
            v: View,
        ) {
            val text = (v as? TextView)?.text
            if (text.isNullOrEmpty()) {
                return
            }
            Repository.get().openUriModel.openUri(context, text.toString())
        }
    }

    companion object {
        fun ofServer(
            context: Context,
            server: MediaServer,
        ): PropertyAdapter = ServerPropertyAdapter(context, server)

        fun ofContent(
            context: Context,
            entity: ContentEntity,
        ): PropertyAdapter = ContentPropertyAdapter(context, entity)

        private val URL_PATTERN = Pattern.compile("https?://[\\w/:%#$&?()~.=+\\-]+")

        private fun isSpace(
            c: Char,
        ): Boolean = c <= '\u0020' || c == '\u00A0' || c == '\u3000'
    }
}
